File: easy/001-100/027/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use time::util::is_leap_year;

fn main() {
    println!("rad");
}

fn determine_year_data(year: i32) -> (i32, bool) {
    (((year as f32 / 100.0).ceil() as i32), is_leap_year(year))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_determine_year_data() {
        assert_eq!(determine_year_data(2000), (20, true));
        assert_eq!(determine_year_data(1996), (20, true));
        assert_eq!(determine_year_data(1900), (19, true));
    }
}

File: easy/001-100/027/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"math"
)

func main() {
	for _, year := range []int{1996, 1900} {
		fmt.Printf("\nEnter Year: %d\n", year)
		fmt.Printf("Century: %d\n", getCentury(year))
		fmt.Printf("Leap Year: ")
		if isLeapYear(year) {
			fmt.Printf("Yes\n")
		} else {
			fmt.Printf("No\n")
		}
	}
}

// Determines whether or not the given year is a leap year
// https://en.wikipedia.org/wiki/Leap_year#Algorithm
func isLeapYear(year int) bool {
	if 0 != year%4 {
		return false
	} else if 0 != year%100 {
		return true
	} else if 0 != year%400 {
		return false
	}
	return true
}

// Gets the century for the given year
// Centuries begin at 1, not 0
func getCentury(year int) int {
	return int(math.Floor(float64(year-1)/100)) + 1
}
